NBitcoin/Protocol/NetworkAddress.cs

```diff
@@ -5,113 +5,383 @@
+		/// Used when (un)serializing addresses in ADDRv1 format (pre-BIP155).
+		static readonly byte[] IPV4_IN_IPV6_PREFIX = new byte[]{
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xFF
+		};
+
+		/// Prefix of an IPv6 address when it contains an embedded TORv2 address.
+		/// Used when (un)serializing addresses in ADDRv1 format (pre-BIP155).
+		/// Such dummy IPv6 addresses are guaranteed to not be publicly routable as they
+		/// fall under RFC4193's fc00::/7 subnet allocated to unique-local addresses.
+		static readonly byte[] TORV2_IN_IPV6_PREFIX = new byte[]{
+			0xFD, 0x87, 0xD8, 0x7E, 0xEB, 0x43
+		};
+
+		/// Prefix of an IPv6 address when it contains an embedded "internal" address.
+		/// Used when (un)serializing addresses in ADDRv1 format (pre-BIP155).
+		/// The prefix comes from 0xFD + SHA256("bitcoin")[0:5].
+		/// Such dummy IPv6 addresses are guaranteed to not be publicly routable as they
+		/// fall under RFC4193's fc00::/7 subnet allocated to unique-local addresses.
+		static readonly byte[] INTERNAL_IN_IPV6_PREFIX = new byte[]{
+			0xFD, 0x6B, 0x88, 0xC0, 0x87, 0x24 // 0xFD + sha256("bitcoin")[0:5].
+		};
+
+
+		static readonly byte[] IPV6_NONE = new byte[ADDR_IPV6_SIZE];
+
+		private byte network = (byte)Network.IPv6;
+		private byte[] addr = new byte[ADDR_IPV6_SIZE];
+
+		private ushort port;
+
+		public ushort Port => port;
+
+		static uint TIME_INIT = 100000000;
+
+		// disk and network only
+		internal uint nTime = TIME_INIT;
+
+		private ulong services = (ulong)NodeServices.Nothing;
+
 		public NetworkAddress()
 		{
+		}
 
+		public NetworkAddress(IPAddress ip, int port)
+			: this(ip)
+		{
+			this.port = (ushort)port;
 		}
-		public NetworkAddress(IPEndPoint endpoint)
+
+		public NetworkAddress(IPAddress ip)
 		{
-			Endpoint = endpoint;
+			SetIp(ip);
 		}
-		public NetworkAddress(IPAddress address, int port)
+
+		public NetworkAddress(EndPoint endPoint)
 		{
-			Endpoint = new IPEndPoint(address, port);
+			Endpoint = endPoint;;
 		}
 
-		internal uint ntime;
-		ulong service = 1;
-		byte[] ip = new byte[16];
-		ushort port;
+		public bool IsIPv4 => (Network)network == Network.IPv4;
+		public bool IsIPv6 => (Network)network == Network.IPv6;
+		public bool IsTor => (Network)network == Network.Onion;
+		public bool IsI2P => (Network)network == Network.I2P;
+		public bool IsCjdns => (Network)network == Network.Cjdns;
 
-		public ulong Service
+		public EndPoint Endpoint
 		{
 			get
 			{
-				return service;
+				if (IsAddrV1Compatible)
+				{
+					return new IPEndPoint(new IPAddress(SerializeV1Array()), port);
+				}
+				else
+				{
+					return new DnsEndPoint(ToAddressString(), port);
+				}
 			}
 			set
 			{
-				service = value;
+				if (value is IPEndPoint ipEndPoint)
+				{
+					SetIp(ipEndPoint.Address);
+					port = (ushort)ipEndPoint.Port;
+				}
+				else if (value is DnsEndPoint dnsEndPoint)
+				{
+					SetSpecial(dnsEndPoint.Host);

```

Comment:
What if is not an onion address? Should you check return value of `SetSpecial` to avoid setting port if it is NOT an onion address?

Reply:
You're right, `SetSpecial` was ported from Bitcoin Core and it seems it its first intention was to set addresses that don't fit into IP addresses like Tor v3 and I2P but in the end it is used only for Tor v3. I think that's an error in core and here. I should create a PR in core first.